spring-pulsar/src/main/java/org/springframework/pulsar/config/ListenerEndpointRegistry.java

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.pulsar.config;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.BeanInitializationException;
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+import org.springframework.context.ApplicationListener;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.context.SmartLifecycle;
+import org.springframework.context.event.ContextRefreshedEvent;
+import org.springframework.lang.Nullable;
+import org.springframework.pulsar.listener.ListenerContainerRegistry;
+import org.springframework.pulsar.listener.MessageListenerContainer;
+import org.springframework.util.Assert;
+
+/**
+ * Creates the necessary container instances for the registered
+ * {@linkplain ListenerEndpoint endpoints}. Also manages the lifecycle of the listener
+ * containers, in particular within the lifecycle of the application context.
+ *
+ * <p>
+ * Contrary to containers created manually, listener containers managed by registry are
+ * not beans in the application context and are not candidates for autowiring. Use
+ * {@link #getListenerContainers()} if you need to access this registry's listener
+ * containers for management purposes. If you need to access to a specific message
+ * listener container, use {@link #getListenerContainer(String)} with the id of the
+ * endpoint.
+ *
+ * @param <C> listener container type.
+ * @param <E> listener endpoint type.
+ * @author Soby Chacko
+ * @author Christophe Bornet
+ */
+public class ListenerEndpointRegistry<C extends MessageListenerContainer, E extends ListenerEndpoint<C>>
+		implements ListenerContainerRegistry, DisposableBean, SmartLifecycle, ApplicationContextAware,
+		ApplicationListener<ContextRefreshedEvent> {
+
+	private final Class<? extends C> type;
```

Comment:
**[For Later]** A more descriptive name may improve readability where this is used below.